File: OpenTkCSharp/WinFormsOpenGL31/Chapter007/015_ColoredCube/ShaderProgram.cs
using System;
using OpenTK.Graphics.OpenGL;
using System.Windows.Forms;
using System.IO;

/// <summary>
/// Allows to get a shader program ID
/// </summary>
class ShaderProgram
{
    public static int InitializeAndGetID(
        string vertexShaderPath,
        string fragmentShaderPath)
    {
        int vShader = GetShaderID(vertexShaderPath, ShaderType.VertexShader);
        int fShader = GetShaderID(fragmentShaderPath, ShaderType.FragmentShader);

        int program = GL.CreateProgram();
        GL.AttachShader(program, vShader);
        GL.AttachShader(program, fShader);
        GL.LinkProgram(program);
        GL.UseProgram(program);

        return program;
    }
    
    private static int GetShaderID(string shaderPath, ShaderType shaderType)
    {
        string shaderSource;
        try
        {
            shaderSource = File.ReadAllText(shaderPath);
        }
        catch (Exception)
        {
            return -1;
        }

        int shader = GL.CreateShader(shaderType);

        GL.ShaderSource(shader, shaderSource);
        GL.CompileShader(shader);

        // Check the compilation status
        int ok;
        GL.GetShader(shader, ShaderParameter.CompileStatus, out ok);
        if (ok == 0)
        {
            string shaderInfo = GL.GetShaderInfoLog(shader);
            MessageBox.Show(string.Format("Failed to compile the shader '{0}'. Message: {1}",
                shaderType.ToString(), shaderInfo));
            return -1;
        }

        return shader;
    }
}

File: glMatrixTypeScript/ch07/015_colored-cube/src/ShaderProgram.ts

export class ShaderProgram
{
    public static InitializeAndGetID(
        gl: WebGLRenderingContext,
        vertexShaderElementName: string,
        fragmentShaderElementName: string): WebGLProgram
    {
        let program = gl.createProgram();

        let vShader = this.GetShaderID(gl, vertexShaderElementName, gl.VERTEX_SHADER);
        let fShader = this.GetShaderID(gl, fragmentShaderElementName, gl.FRAGMENT_SHADER);
        
        gl.attachShader(program, vShader);
        gl.attachShader(program, fShader);
        gl.linkProgram(program);
        gl.useProgram(program);

        return program;
    }

    private static GetShaderID(
        gl: WebGLRenderingContext,
        shaderElementName: string,
        shaderType: number): WebGLShader
    {
        let shaderElement = document.getElementById(shaderElementName);
        let shaderSource = shaderElement.firstChild.textContent;
        let shader = gl.createShader(shaderType);

        gl.shaderSource(shader, shaderSource);
        gl.compileShader(shader);

        let ok = gl.getShaderParameter(shader, gl.COMPILE_STATUS);
        if (!ok)
        {
            let shaderInfo = gl.getShaderInfoLog(shader);
            console.log("Failed to compile the shader. Message: " + shaderInfo);
            return null;
        }

        return shader;
    }
}